unified-scheduler-logic/src/lib.rs

```diff
@@ -1,20 +1,1333 @@
+#[derive(Debug)]
+struct PageInner {
+    usage: PageUsage,
+    blocked_tasks: VecDeque<(Task, RequestedUsage)>,
+}
+
+impl Default for PageInner {
+    fn default() -> Self {
+        Self {
+            usage: PageUsage::default(),
+            blocked_tasks: VecDeque::with_capacity(1024),
+        }
+    }
+}
+
+impl PageInner {
+    fn push_blocked_task(&mut self, task: Task, requested_usage: RequestedUsage) {
+        self.blocked_tasks.push_back((task, requested_usage));
+    }
+
+    fn has_no_blocked_task(&self) -> bool {
+        self.blocked_tasks.is_empty()
+    }
+
+    #[must_use]
+    fn pop_unblocked_next_task(&mut self) -> Option<(Task, RequestedUsage)> {
+        self.blocked_tasks.pop_front()
+    }
+
+    #[must_use]
+    fn blocked_next_task(&self) -> Option<(&Task, RequestedUsage)> {
+        self.blocked_tasks
+            .front()
+            .map(|(task, requested_usage)| (task, *requested_usage))
+    }
+
+    #[must_use]
+    fn pop_blocked_next_readonly_task(&mut self) -> Option<(Task, RequestedUsage)> {
+        if matches!(
+            self.blocked_next_task(),
+            Some((_, RequestedUsage::Readonly))
+        ) {
+            self.pop_unblocked_next_task()
+        } else {
+            None
+        }
+    }
+}
+
+const_assert_eq!(mem::size_of::<TokenCell<PageInner>>(), 40);
+
+/// Scheduler's internal data for each address ([`Pubkey`](`solana_sdk::pubkey::Pubkey`)). Very
+/// opaque wrapper type; no methods just with [`::clone()`](Clone::clone) and
+/// [`::default()`](Default::default).
+#[derive(Debug, Clone, Default)]
+pub struct Page(Arc<TokenCell<PageInner>>);
+const_assert_eq!(mem::size_of::<Page>(), 8);
+
+/// A high-level `struct`, managing the overall scheduling of [tasks](Task), to be used by
+/// `solana-unified-scheduler-pool`.
+pub struct SchedulingStateMachine {
+    last_task_index: Option<usize>,
+    unblocked_task_queue: VecDeque<Task>,
+    active_task_count: ShortCounter,
+    handled_task_count: ShortCounter,
+    unblocked_task_count: ShortCounter,
+    total_task_count: ShortCounter,
+    count_token: BlockedPageCountToken,
+    page_token: PageToken,
+}
+const_assert_eq!(mem::size_of::<SchedulingStateMachine>(), 64);
+
+impl SchedulingStateMachine {
+    pub fn has_no_active_task(&self) -> bool {
+        self.active_task_count.is_zero()
+    }
+
+    pub fn unblocked_task_queue_count(&self) -> usize {
+        self.unblocked_task_queue.len()
+    }
+
+    pub fn active_task_count(&self) -> u32 {
+        self.active_task_count.current()
+    }
+
+    pub fn handled_task_count(&self) -> u32 {
+        self.handled_task_count.current()
+    }
+
+    pub fn unblocked_task_count(&self) -> u32 {
+        self.unblocked_task_count.current()
+    }
+
+    pub fn total_task_count(&self) -> u32 {
+        self.total_task_count.current()
+    }
+
+    #[must_use]
+    pub fn schedule_task(&mut self, task: Task) -> Option<Task> {
```

Comment:
Additionally it's unclear to me why we want to return the task at all here instead of just returning something a little bit more direct such as a bool indicating it can be immediately sent as a payload. since we do not modify the task in case of success, that seems fine?

Maybe you have plans to do some modifications to Task in here in the future?

Reply:
> Maybe you have plans to do some modifications to Task in here in the future?

yeah. I'm planning to reuse `Arc::strong_count()` instead of `blocked_task_count()`. I'll update the docs.